File: util-parser/src/main/scala/io/github/kory33/guardedqueries/parser/FormulaParsers.java
package io.github.kory33.guardedqueries.parser;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import io.github.kory33.guardedqueries.core.utils.extensions.SetLikeExtensions;
import io.github.kory33.guardedqueries.core.utils.extensions.VariableSetExtensions;
import org.javafp.data.IList;
import org.javafp.parsecj.Combinators;
import org.javafp.parsecj.Parser;
import org.javafp.parsecj.Text;
import org.javafp.parsecj.input.Input;
import uk.ac.ox.cs.gsat.GTGD;
import uk.ac.ox.cs.pdq.fol.*;

import java.util.Arrays;
import java.util.HashSet;

public class FormulaParsers {
    private FormulaParsers() {
    }

    // here is the grammar for formulae that we will deal with
    // (we preprocess all input strings and remove all spaces)
    //
    // term ::= <nonempty sequence of characters not including ',' nor ')'>
    // predicate ::= <nonempty sequence of characters not including '(' and not starting with 'EE' or '->'>
    // variable ::= <nonempty sequence of characters not including ',' or '.' which
    //               the parsing context does not judge as constant>
    //
    // terms ::= term | term ',' terms
    // atom ::= predicate '()' | predicate '(' terms ')'
    // atoms ::= atom | atom ',' atoms
    // variables ::= variable | variable ',' variables

    public static Parser<Character, Term> termParser(FormulaParsingContext context) {
        final var termChars = Combinators.<Character>satisfy(c -> c != ',' && c != ')');
        return termChars.many1().map(chars -> {
            final var termString = IList.listToString(chars);

            if (context.isConstantSymbol().test(termString)) {
                return context.mapConstantSymbolToConstant().apply(termString);
            } else {
                return Variable.create(termString);
            }
        });
    }

    public static final Parser<Character, String> predicateNameParser =
            Combinators.<Character>satisfy(c -> c != '(').many1().bind(chars -> {
                final var predicateString = IList.listToString(chars);

                if (predicateString.startsWith("EE") || predicateString.startsWith("->")) {
                    return Combinators.fail("predicate name cannot start with 'EE' or '->', found: " + predicateString);
                } else {
                    return Combinators.retn(predicateString);
                }
            });

    public static Parser<Character, Atom> atomParser(FormulaParsingContext context) {
        return predicateNameParser.bind(predicateName ->
                Text.chr('(').then(termParser(context).sepBy(Text.chr(','))).bind(terms ->
                        Text.chr(')').map(_ignored -> {
                            final var arity = terms.size();
                            final var predicate = Predicate.create(predicateName, arity);
                            return Atom.create(predicate, terms.stream().toArray(Term[]::new));
                        })
                )
        );
    }

    public static Parser<Character, Variable> variableParser(FormulaParsingContext context) {
        return Combinators
                .<Character>satisfy(c -> c != ',' && c != '.')
                .many1()
                .bind(chars -> {
                    final var variableString = IList.listToString(chars);
                    if (context.isConstantSymbol().test(variableString)) {
                        return Combinators
                                .fail("a constant symbol " + variableString + " cannot be bound in quantifier");
                    } else {
                        return Combinators.retn(Variable.create(variableString));
                    }
                });
    }

    public static Parser<Character, IList<Variable>> existentialVariableBindingsParser(FormulaParsingContext context) {
        return Text
                .string("EE")
                .then(variableParser(context).sepBy(Text.chr(',')))
                .bind(variables -> Text.chr('.').bind(_ignored -> {
                    final var variableList = ImmutableList.copyOf(variables.stream().iterator());

                    if (ImmutableSet.copyOf(variableList).size() != variableList.size()) {
                        return Combinators.fail("existential variables must be distinct, found " + variableList);
                    } else {
                        return Combinators.retn(variables);
                    }
                }));
    }

    // We use the following grammar for TGDs
    //
    // gtgdHead ::= atoms | 'EE' variables '.' atoms
    // gtgd ::= atoms '->' gtgdHead
    public static Parser<Character, TGD> tgdParser(FormulaParsingContext context) {
        record TGDHead(IList<Atom> headAtoms, IList<Variable> existentialVariables) {
        }

        final var existentialHeadParser = existentialVariableBindingsParser(context)
                .bind(existentialVariables ->
                        atomParser(context).sepBy(Text.chr(',')).map(headAtoms ->
                                new TGDHead(headAtoms, existentialVariables)
                        )
                );

        final var nonExistentialHeadParser = atomParser(context).sepBy(Text.chr(',')).map(headAtoms ->
                new TGDHead(headAtoms, IList.empty())
        );

        final var headParser = existentialHeadParser.or(nonExistentialHeadParser);

        return atomParser(context).sepBy(Text.chr(',')).bind(bodyAtoms -> Text.string("->").then(headParser).bind(head -> {
            final var bodyAtomsArray = bodyAtoms.stream().toArray(Atom[]::new);
            final var headAtoms = head.headAtoms.stream().toArray(Atom[]::new);
            final var tgd = TGD.create(bodyAtomsArray, headAtoms);

            final var declaredExistentialVariables = VariableSetExtensions.sortBySymbol(head.existentialVariables.toList());
            final var foundExistentialVariables = VariableSetExtensions.sortBySymbol(Arrays.asList(tgd.getExistential()));

            if (!foundExistentialVariables.equals(declaredExistentialVariables)) {
                return Combinators.fail("existential variables in head do not match those in body: " +
                        "declared: " + declaredExistentialVariables + ", found: " + foundExistentialVariables);
            } else {
                return Combinators.retn(tgd);
            }
        }));
    }

    public static Parser<Character, GTGD> gtgdParser(FormulaParsingContext context) {
        return tgdParser(context).map(tgd -> {
            final var bodyAtoms = new HashSet<>(Arrays.asList(tgd.getBodyAtoms()));
            final var headAtoms = new HashSet<>(Arrays.asList(tgd.getHeadAtoms()));
            return new GTGD(bodyAtoms, headAtoms);
        });
    }

    public static Parser<Character, ConjunctiveQuery> conjunctiveQueryParser(FormulaParsingContext context) {
        final var existentialQueryParser = existentialVariableBindingsParser(context).bind(existentialVariables ->
                atomParser(context).sepBy(Text.chr(',')).map(queryAtoms -> {
                    final var atomVariables = queryAtoms.stream()
                            .flatMap(atom -> Arrays.stream(atom.getFreeVariables()))
                            .toList();

                    final var freeVariables = SetLikeExtensions.difference(
                            atomVariables,
                            existentialVariables.toList()
                    ).toArray(Variable[]::new);

                    final var queryAtomsArray = queryAtoms.stream().toArray(Atom[]::new);

                    return ConjunctiveQuery.create(freeVariables, queryAtomsArray);
                })
        );

        final var nonExistentialQueryParser = atomParser(context).sepBy(Text.chr(',')).map(queryAtoms -> {
            final var atomVariables = queryAtoms.stream()
                    .flatMap(atom -> Arrays.stream(atom.getFreeVariables()))
                    .toList();
            final var freeVariables = atomVariables.toArray(Variable[]::new);
            final var queryAtomsArray = queryAtoms.stream().toArray(Atom[]::new);
            return ConjunctiveQuery.create(freeVariables, queryAtomsArray);
        });

        return existentialQueryParser.or(nonExistentialQueryParser);
    }

    public record WhitespaceIgnoringParser<FormulaClass>(Parser<Character, FormulaClass> parser) {
        public FormulaClass parse(String formulaString) {
            final var spaceRemoved = formulaString.replace(" ", "");
            try {
                return parser.parse(Input.of(spaceRemoved)).getResult();
            } catch (Exception e) {
                throw new IllegalArgumentException("invalid formula string: " + formulaString, e);
            }
        }
    }

    public record GTGDParser(FormulaParsingContext context) {
        public GTGD parse(String formulaString) {
            final var spaceRemoved = formulaString.replace(" ", "");
            try {
                return gtgdParser(context).parse(Input.of(spaceRemoved)).getResult();
            } catch (Exception e) {
                throw new IllegalArgumentException("invalid GTGD string: " + formulaString, e);
            }
        }
    }

    public record ConjunctiveQueryParser(FormulaParsingContext context) {
        public ConjunctiveQuery parse(String formulaString) {
            final var spaceRemoved = formulaString.replace(" ", "");
            try {
                return conjunctiveQueryParser(context).parse(Input.of(spaceRemoved)).getResult();
            } catch (Exception e) {
                throw new IllegalArgumentException("invalid conjunctive query string: " + formulaString, e);
            }
        }
    }
}

File: formula-parsers/src/main/scala/io/github/kory33/guardedqueries/parser/FormulaParsers.scala
package io.github.kory33.guardedqueries.parser

import uk.ac.ox.cs.gsat.GTGD
import uk.ac.ox.cs.pdq.fol.*

import scala.util.parsing.combinator.*

object FormulaParsers extends RegexParsers {
  // here is the grammar for formulae that we will deal with
  // (we preprocess all input strings and remove all spaces)
  //
  // term ::= <nonempty sequence of characters not including ',' nor ')'>
  // predicate ::= <nonempty sequence of characters not including '(' and not starting with 'EE' or '->'>
  // variable ::= <nonempty sequence of characters not including ',' or '.' which
  //               the parsing context does not judge as constant>
  //
  // terms ::= term | term ',' terms
  // atom ::= predicate '()' | predicate '(' terms ')'
  // atoms ::= atom | atom ',' atoms
  // variables ::= variable | variable ',' variables

  private def term(using context: FormulaParsingContext): Parser[Term] = {
    regex("""[^,)]+""".r).map(termString =>
      if (context.isConstantSymbol(termString)) context.mapConstantSymbolToConstant(termString)
      else Variable.create(termString)
    )
  }

  private val predicate: Parser[String] = {
    for {
      predicateString <- regex("""[^(]+""".r)
      _ <-
        if (predicateString.startsWith("EE") || predicateString.startsWith("->"))
          failure(s"predicate name cannot start with 'EE' or '->', found: $predicateString")
        else
          success(())
    } yield predicateString
  }

  private def atom(using FormulaParsingContext): Parser[Atom] = {
    for {
      predicateName <- predicate
      terms <- literal("(") ~> repsep(term, literal(",")) <~ literal(")")
    } yield {
      val predicate = Predicate.create(predicateName, terms.size)
      Atom.create(predicate, terms.toArray: _*)
    }
  }

  private def existentialVariableBindingsParser(
    using context: FormulaParsingContext
  ): Parser[List[Variable]] = {
    val variableBoundInQuantifier: Parser[Variable] = {
      """[^,.]+""".r.flatMap(variableString =>
        if (context.isConstantSymbol(variableString))
          failure(s"a constant symbol $variableString cannot be bound in quantifier")
        else
          success(Variable.create(variableString))
      )
    }

    for {
      variables <-
        literal("EE") ~> repsep(variableBoundInQuantifier, ",") <~ literal(".")
      _ <-
        if (variables.distinct.size != variables.size)
          failure(s"existential variables must be distinct, found $variables")
        else
          success(())
    } yield variables
  }

  // We use the following grammar for TGDs
  //
  // gtgdHead ::= atoms | 'EE' variables '.' atoms
  // gtgd ::= atoms '->' gtgdHead

  private def tgdParser(using FormulaParsingContext): Parser[TGD] = {
    case class TGDHead(headAtoms: List[Atom], existentialVariables: List[Variable])

    val atoms = repsep(atom, literal(","))

    val existentialHeadParser =
      for {
        existentialVariables <- existentialVariableBindingsParser
        headAtoms <- atoms
      } yield TGDHead(headAtoms, existentialVariables)

    val nonExistentialHeadParser =
      for {
        headAtoms <- atoms
      } yield TGDHead(headAtoms, List.empty)

    val headParser = existentialHeadParser | nonExistentialHeadParser

    for {
      bodyAtoms <- atoms
      _ <- literal("->")
      head <- headParser

      tgd = TGD.create(bodyAtoms.toArray, head.headAtoms.toArray)
      declaredExistentialVariables = head.existentialVariables.toSet
      foundExistentialVariables = tgd.getExistential.toSet
      _ <-
        if (foundExistentialVariables != declaredExistentialVariables)
          failure(
            "existential variables in head do not match those in body: " +
              s"declared: $declaredExistentialVariables, found: $foundExistentialVariables"
          )
        else
          success(())
    } yield tgd
  }

  def gtgdParser(using FormulaParsingContext): Parser[GTGD] = {
    tgdParser.map { tgd =>
      import scala.jdk.CollectionConverters.*

      GTGD(
        tgd.getBodyAtoms.toSet.asJava,
        tgd.getHeadAtoms.toSet.asJava
      )
    }
  }

  def conjunctiveQueryParser(using FormulaParsingContext): Parser[ConjunctiveQuery] = {
    val existentialQueryParser =
      for {
        existentialVariables <- existentialVariableBindingsParser
        queryAtoms <- repsep(atom, literal(","))
      } yield {
        val atomVariables = queryAtoms.flatMap(_.getFreeVariables).toSet
        val freeVariables = atomVariables -- existentialVariables

        ConjunctiveQuery.create(freeVariables.toArray, queryAtoms.toArray)
      }

    val nonExistentialQueryParser =
      for {
        queryAtoms <- repsep(atom, literal(","))
      } yield {
        val variablesInAtoms = queryAtoms.flatMap(_.getFreeVariables).toSet
        ConjunctiveQuery.create(variablesInAtoms.toArray, queryAtoms.toArray)
      }

    existentialQueryParser | nonExistentialQueryParser
  }

  case class IgnoreWhitespaces[FormulaClass](parser: Parser[FormulaClass]) {
    def parse(formulaString: String): FormulaClass = {
      val spaceRemoved = formulaString.replace(" ", "")

      FormulaParsers.parse(parser, spaceRemoved) match {
        case Success(result, _) => result
        case n: NoSuccess =>
          throw new IllegalArgumentException(
            s"invalid formula string: $formulaString, error: ${n.msg}"
          )
      }
    }
  }
}
